src/file_patcher.rs

```diff
@@ -30,100 +28,36 @@ impl<'a> FilePatcher<'a> {
                 return Ok(None);
             }
             let line = line.unwrap();
-            let line_patcher = LinePatcher::new(&line);
-            let new_line = line_patcher.replace(&query);
-            if new_line != line {
-                let replacement = Replacement {
-                    line_no: num + 1,
-                    old: line,
-                    new: new_line.to_string(),
-                    query: query.description(),
-                };
-                replacements.push(replacement);
-                new_contents.push_str(&new_line);
-            } else {
-                new_contents.push_str(&line);
+            let replacement = replace(&line, &query);
+            match replacement {
+                None => new_contents.push_str(&line),
+                Some(replacement) => {
+                    num_replacements += 1;
+                    let lineno = num + 1;
+                    let prefix = format!("{}:{} ", path.to_string_lossy(), lineno);
+                    let new_line = replacement.output();
+                    replacement.print_self(&prefix);
+                    new_contents.push_str(&new_line);
```

Comment:
I think it makes perfect sense for FilePatch - but not as much for the main replace function.

Reply:
Oh yeah definitely, I think code should be as simple as possible if it's not actively visibly slow. I just notice these things!

Reply:
So ... I tried but fail to use a Vec<Cow<str>> inside the FilePatcher.

But I could not get it to work. I'm not sure we can make the borrow checker happy, with for loop with the `chunk` and `line` variables ...

Reply:
nice tricks!!! thx :)

Reply:
Oh, you're right! I got confused, ruplacer prints the diffs directly, but it builds the output and writes it whole at the end. In my mind everything was streaming, but you're completely right my idea doesn't make sense here since we read the input line by line but FilePatcher's output is the whole document, the lines can't live long enough.

A more extreme architecture would be to make FilePatcher itself be fully an Iterator, and each time you call next() it would read 1 line in the input and print 1 line in the output. That's an even more crazy design, because now you can process documents as big as you want. But it's probably a lot more work and not really necessary!

See, I'm really overzealous with the Cows, can't use em everywhere it turns out!

Reply:
> That's an even more crazy design, because now you can process documents as big as you want.

That would work if we were reading one file and writing another, but that's not what we're doing here :)

Reply:
True!